server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java

```diff
@@ -0,0 +1,216 @@
+          "Could not read root metadata from ZooKeeper due to interrupt", ex);
+    } catch (KeeperException ex) {
+      throw new IllegalStateException(
+          "Could not read root metadata from ZooKeeper because of ZooKeeper exception", ex);
+    }
+  }
+
+  @Override
+  public void upgradeRoot(@NonNull ServerContext context) {
+    log.debug("Upgrade root: upgrading to data version {}", METADATA_FILE_JSON_ENCODING);
+    var rootName = Ample.DataLevel.METADATA.metaTable();
+    processReferences(context, rootName);
+  }
+
+  @Override
+  public void upgradeMetadata(@NonNull ServerContext context) {
+    log.debug("Upgrade metadata: upgrading to data version {}", METADATA_FILE_JSON_ENCODING);
+    var metaName = Ample.DataLevel.USER.metaTable();
+    processReferences(context, metaName);
+  }
+
+  private void processReferences(ServerContext context, String tableName) {
+    // not using ample to avoid StoredTabletFile because old file ref is incompatible
+    try (AccumuloClient c = Accumulo.newClient().from(context.getProperties()).build();
+        BatchWriter batchWriter = c.createBatchWriter(tableName); Scanner scanner =
+            new IsolatedScanner(context.createScanner(tableName, Authorizations.EMPTY))) {
+
+      scanner.fetchColumnFamily(DataFileColumnFamily.NAME);
+      scanner.fetchColumnFamily(ChoppedColumnFamily.NAME);
+      scanner.fetchColumnFamily(ExternalCompactionColumnFamily.NAME);
+      scanner.forEach((k, v) -> {
+        var family = k.getColumnFamily();
+        if (family.equals(DataFileColumnFamily.NAME)) {
+          upgradeDataFileCF(k, v, batchWriter, tableName);
+        } else if (family.equals(ChoppedColumnFamily.NAME)) {
+          removeChoppedCF(k, batchWriter, tableName);
+        } else if (family.equals(ExternalCompactionColumnFamily.NAME)) {
+          removeExternalCompactionCF(k, batchWriter, tableName);
+        } else {
+          throw new IllegalStateException("Processing: " + tableName
+              + " Received unexpected column family processing references: " + family);
+        }
+      });
+    } catch (MutationsRejectedException mex) {
+      log.warn("Failed to update reference for table: " + tableName);
+      log.warn("Constraint violations: {}", mex.getConstraintViolationSummaries());
+      throw new IllegalStateException("Failed to process table: " + tableName, mex);
+    } catch (Exception ex) {
+      throw new IllegalStateException("Failed to process table: " + tableName, ex);
+    }
+  }
+
+  @VisibleForTesting
+  void upgradeDataFileCF(final Key key, final Value value, final BatchWriter batchWriter,
+      final String tableName) {
+    String file = key.getColumnQualifier().toString();
+    // filter out references if they are in the correct format already.
+    if (fileNeedsConversion(file)) {
+      var fileJson = StoredTabletFile.of(new Path(file)).getMetadataText();
+      try {
+        Mutation update = new Mutation(key.getRow());
```

Comment:
It would be nice to generate a single mutation for all the tablets files.  That may be faster and would result in atomically updating the tablets files. However I don't think this has to be done.  If it were done, it could be in a follow on PR. What is here seems correct and is probably fast enough for the one time upgrade case.

Reply:
I worried about memory usage for a large number of files as a single mutation - but this would also eliminate the need to check for the file needs conversion issue comment on above.

The `fileNeedsConversion` check was added to handle partial upgrades where some files had been converted and then an error would halt the upgrade leaving some files converted, some not.  The halt was trigger by errors during development, but seemed that it could allow an admin to perform corrective action and the "restart" the upgrade if it happened in general.

Reply:
For the upgrade process, there is no difference if the upgrade code were to fail on mutation updates if it was for a mutation within a tablet or for a whole tablet mutation - that is, there should be no impact to the upgrade code being re-run with a partial upgrade that had previously completed at any stage.  

Being that the upgrade code is running before things are brought online, any inconsistency in file representations needs to be resolved before the next level can be brought online.  For example, until all root table file references pointing to the metadata tablet files are resolved, the upgrade will not progress to the point of processing the metadata file references.  This is independent if the inconsistencies were within a tablet of for any tablet in general.